### PR TITLE
Improve Level API a bit

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of https://github.com/rinja-rs/rinja/issues/30.

It already improves the situation a bit. Not sure if the rest is worth improving or not though.

### CODE REVIEW

File: rinja_parser/src/lib.rs [hunk 1]
```diff
@@ -635,14 +635,17 @@ impl<'a> State<'a> {
         }
     }
 
-    fn nest<'b>(&self, i: &'b str) -> ParseResult<'b, ()> {
+    fn nest<'b, T, F: FnMut(&'b str) -> ParseResult<'b, T>>(
+        &self,
+        i: &'b str,
+        mut callback: F,
```
Comment (Kijewski) [inline rinja_parser/src/lib.rs:641]: `F` can be `FnOnce`. Currently that does not matter, but maybe we refactor the parser code later.

File: rinja_parser/src/lib.rs [hunk 2]
```diff
@@ -635,14 +635,17 @@ impl<'a> State<'a> {
...
+    ) -> ParseResult<'b, T> {
+        let prev_level = self.level.get();
         let (_, level) = self.level.get().nest(i)?;
```
Comment (Kijewski) [inline rinja_parser/src/lib.rs:644]: You could directly call `prev_level.nest(i)`.

File: rinja_parser/src/node.rs [hunk 3]
```diff
@@ -90,10 +90,7 @@ impl<'a> Node<'a> {
             }
         };
 
-        let (i, _) = s.nest(j)?;
-        let result = func(i, s);
-        s.leave();
-        let (i, node) = result?;
```
Comment (Kijewski) [inline rinja_parser/src/node.rs:96]: It is much clearer what is going on with you change, and the code does not much more `nom`-like. Clear advancement!

**Review by Kijewski — CHANGES_REQUESTED** [review]

File: rinja_parser/src/lib.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply rinja_parser/src/lib.rs:641]: Good point!

File: rinja_parser/src/lib.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply rinja_parser/src/lib.rs]: Indeed!

**Comment by GuillaumeGomez** [comment]
There is still the original `Level`. I'm still wondering if `nest` should return a non-`Copy` type which would contain a mutable reference to the calling `Level` or something along the line. Really not clear about it...

**Comment by GuillaumeGomez** [comment]
Fixed CI. :)

**Review by Kijewski — APPROVED** [review]

**Pull request merged**